### DISCUSSION TITLE
How to annotate and add Markers to Smith Chart output?

### DISCUSSION THREAD

**Discussion by shabaz123** [scikit-rf/scikit-rf#1039] [Q&A] [answered]
Hello,

I've been using scikit-rf and it is great for plotting Smith charts! Here's an example:
<img width="555" alt="smith-chart-example" src="https://github.com/scikit-rf/scikit-rf/assets/6280966/38395c8d-3a9d-416f-87c7-94b610dcba1e">
The code I used to draw that is here: https://gist.github.com/shabaz123/0f9ad5392812d55db9dcefe48f5782bb

However, I had two questions.

Firstly, has anyone got any example code that demonstrates a large Smith chart, with axes with numbers on them, such as this one:
<img width="785" alt="smith-chart-desired" src="https://github.com/scikit-rf/scikit-rf/assets/6280966/327dab99-617e-4022-91a0-c13e1da20dc2">
I sometimes like the simplicity of the plain Smith chart, but sometimes really wish to have the axes marked. I was thinking, perhaps there is a way to 'overlay' a png image of a detailed Smith chart and then draw on top with scikit-rf and matplotlib? My knowledge of these Python libraries is not high : ( 
EDIT: I've discovered how to do this:
<img width="558" alt="smith-chart-example2" src="https://github.com/scikit-rf/scikit-rf/assets/6280966/20570168-7466-43b5-b8cf-470650844711">
That's much better, but it would still be nice if there was a way to overlay the Smith chart on a high-res png image that the user could create with the detail they require.

The second question is, how can markers be drawn on the output traces? For instance, I might need a marker at 1 GHz on the orange trace, and a marker at 2 GHz on the blue trace. 
It would be nice if the markers were interactive (I have googled for that, and apparently pyQT could be used, but it would be great to see a working code example of that, since I can imagine it must be a popular requirement to have interactive markers). 
Even static markers (i.e. set in the code) would be great to see, since currently I'm flying blind when looking at the Smith chart output. I have had to resort to saving to touchstone format, and then using a third-party touchstone-to-smith-chart viewer program, but it's not a great workaround.

Any help/pointers would be gratefully appreciated. Many thanks.

**Comment by mhuser** [comment]
![smith_chart_nice](https://github.com/scikit-rf/scikit-rf/assets/8080934/5cb8b989-2463-4e7f-becf-68b6a0672153)

```Python
import skrf as rf
from matplotlib import pyplot as plt
rf.stylely()

from skrf.data import ring_slot

n = ring_slot.s11
marker_idx = [30, 60, 90]


# prepare figure
fig, axs = plt.subplots(1, 2, figsize=(8, 4))

# nice smith chart
rf.plotting.smith(ax = axs[0], draw_labels = True, ref_imm = 50.0, chart_type = 'z')
n.plot_s_smith(ax = axs[0], linewidth=1.0)
leg1 = axs[0].legend(loc="upper right", fontsize= 6)

# plot markers
col_labels = ['Frequency', 'Real Imag']
row_labels = []
cell_text = []
for i, k in enumerate(marker_idx):
    x = n.s.real[k, 0, 0]
    y = n.s.imag[k, 0, 0]
    f = n.frequency.f_scaled[k]
    row_labels.append(f'M{i}')
    axs[0].scatter(x, y, marker = 'v', s=20, color='k')
    axs[0].annotate(row_labels[-1], (x, y), xytext=(-7, 7), textcoords='offset points', color='k')
    cell_text.append([f'{f} {n.frequency.unit}', f'{x:.4f} + {y:.4f}j'])

# plot the table    
the_table = axs[1].table(cellText=cell_text,
                      colWidths=[0.4] * 2,
                      rowLabels=row_labels,
                      colLabels=col_labels,
                      loc='center')
the_table.auto_set_font_size(False)
the_table.set_fontsize(6)
#the_table.scale(1.5, 1.5)
axs[1].axis("off") 
```

**Comment by mhuser** [comment, marked as answer]
![smith_chart_nice_with_background](https://github.com/scikit-rf/scikit-rf/assets/8080934/21a460ea-f6d0-41df-aadb-e8e1272f98a2)

Hello @shabaz123 thank you for your interest in the smith chart.
I do not know a simple way to have interactive markers but you will find below an example with two traces and your background image.

```Python
import skrf as rf
from matplotlib import pyplot as plt
import numpy as np
rf.stylely()

from skrf.data import ring_slot

n = ring_slot
marker_idx = [30, 60, 90]
colors = ['g', 'r']
to_plot = [0, 1]


# prepare figure
fig, ax = plt.subplots(1, 1, figsize=(7,8))
background = plt.imread('smithchart.png')

# nice smith chart
# tweak background position
ax.imshow(background, extent=[-1.19, 1.14, -1.13, 1.15])
rf.plotting.smith(ax = ax, draw_labels = True, ref_imm = 1.0, chart_type = 'z')

# plot markers
col_labels = ['Frequency', 'Real Imag']
row_labels = []
cell_text = []
for j in to_plot:
    color = colors[j]
    n.plot_s_smith(j, j, ax = ax, linewidth=1.0, color=color)
    for i, k in enumerate(marker_idx):
        x = n.s.real[k, j, j]
        y = n.s.imag[k, j, j]
        z = f'{n.z.real[k, j, j]:.4f} + {n.z.imag[k, j, j]:.4f}j ohm'
        f = n.frequency.f_scaled[k]
        row_labels.append(f'M{i}')
        ax.scatter(x, y, marker = 'v', s=20, color=color)
        ax.annotate(row_labels[-1], (x, y), xytext=(-7, 7), textcoords='offset points', color=color)
        cell_text.append([f'{f} {n.frequency.unit}', z])
leg1 = ax.legend(loc="upper right", fontsize= 6)

# plot the table    
the_table = ax.table(cellText=cell_text,
                      colWidths=[0.4] * 2,
                      rowLabels=row_labels,
                      colLabels=col_labels,
                      rowColours=np.repeat(colors, len(marker_idx)),
                      loc='bottom')
the_table.auto_set_font_size(False)
the_table.set_fontsize(6)
#the_table.scale(1.5, 1.5)
```

**Comment by shabaz123** [comment]
Hi @mhuser 
Wow that's fantastic! Thank you so much - this is brilliant. I'm going to experiment with this now!

**Reply by mhuser** [reply]
Thank you a lot! I will add an example in the documentation for further reference #1041 !
